source/common/http/filter/ip_tagging_filter.cc

```diff
@@ -9,7 +14,34 @@ IpTaggingFilter::~IpTaggingFilter() {}
 
 void IpTaggingFilter::onDestroy() {}
 
-FilterHeadersStatus IpTaggingFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus IpTaggingFilter::decodeHeaders(HeaderMap& headers, bool) {
+  bool is_internal_request =
+      headers.EnvoyInternalRequest() && (headers.EnvoyInternalRequest()->value() ==
+                                         Headers::get().EnvoyInternalRequestValues.True.c_str());
+
+  if ((is_internal_request && config_->requestType() == FilterRequestType::EXTERNAL) ||
+      (!is_internal_request && config_->requestType() == FilterRequestType::INTERNAL) ||
+      (!config_->runtime().snapshot().featureEnabled("ip_tagging.http_filter_enabled", 100))) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  std::vector<std::string> tags =
+      config_->trie().getTags(callbacks_->requestInfo().downstreamRemoteAddress());
+
+  if (!tags.empty()) {
+    const std::string tags_join = absl::StrJoin(tags, ",");
+    Utility::appendToHeader(headers.insertEnvoyIpTags().value(), tags_join);
+
+    // For a large number(ex > 1000) of tags, stats cardinality will be an issue.
+    // If there are use cases with a large set of tags, a way to opt into these stats
+    // should be exposed and other observability options like logging tags need to be implemented.
```

Comment:
Yeah, this seems like something that we're going to have to address at some point beyond low cardinality tagging.